src/dblib_stubs.c

```diff
@@ -30,13 +30,50 @@
 #include <caml/fail.h>
 #include <caml/callback.h>
 #include <caml/custom.h>
+#include <caml/threads.h>
 
 /* OCaml severity.  Keep in sync with OCaml [Dblib]. */
 #define SEVERITY_PROGRAM Val_int(6)
 #define SEVERITY_RESOURCE Val_int(7)
 #define SEVERITY_FATAL Val_int(9)
 #define SEVERITY_CONSISTENCY Val_int(10)
 
+static int release_lock(DBPROCESS *dbproc)
+{
+  int *have_lock = (int*)dbgetuserdata(dbproc);
+
+  // If have_lock wasn't setup, then we're in the msg/error callback for dbopen
+  if (have_lock == NULL) {
```

Comment:
Shouldn't the lock be allocated & initialized when the dbproc value is created (and released by its finalizer)?

Reply:
That's what this pull request does. The problem is that `dbopen` both allocates a dbproc and attempts to connect to the remote server, so if the connection fails, we'll have the exception handler called before we have a chance to call `dbsetuserdata`.